### ISSUE TITLE
Trouble running simv and verilator

### ISSUE THREAD

**Issue by ax-ag16** [chipsalliance/caliptra-rtl#148]
I am two facing two issues:

1. I am able to generate program.hex, iccm.hex, dccm.hex and mailbox.hex but when I run the simv, the simulation hangs indefinitely. I have copied the ecdsa_secp384r1.exe and doe_test_gen.py files from [https://github.com/chipsalliance/caliptra-rtl/issues/120]. I get warnings that are in the attached file below . Please let me know if i am missing any files. I am running this command from CALIPTRA/caliptra-rtl-main/run
[sim.log](https://github.com/chipsalliance/caliptra-rtl/files/11961707/sim.log)

The command has generated empty iccm.hex and mailbox.hex. The dccm.hex files are however different from the ones generated from [https://github.com/chipsalliance/caliptra-rtl/blob/48012ef7c254308052a6d90918abb800d635e1da/src/integration/test_suites/iccm_lock/dccm.hex](url). I have attached the dccm.hex file that's generated from my environment. 
[dccm.txt](https://github.com/chipsalliance/caliptra-rtl/files/11962394/dccm.txt) (uplaoding it is .txt as .hex is not supported )

2. I am not build any verilator related build files after using command `make -C ${CALIPTRA_WORKSPACE}/scratch/agurram/verilator/06302023/run -f ${CALIPTRA_ROOT}/tools/scripts/Makefile TESTNAME=${iccm_lock} debug=1 verilator > verilate.log` 
I have attached the logs :[verilate.log](https://github.com/chipsalliance/caliptra-rtl/files/11961757/verilate.log)
I am running this command from /CALIPTRA/scratch/agurram/verilator/06302023/run

I have also attached the tree to show which directories I'm running from: [tree.log](https://github.com/chipsalliance/caliptra-rtl/files/11961982/tree.log)

My paths for ROOT , WORKSPACE and PATH are  :

export W_ROOT=`pwd`
export PATH=$PATH:/ax_tools/Others/toolchain/riscv/bin
export PATH=$PATH:/usr/bin
export PATH=$PATH:/ax_tools/Others/gcc/8.2.0/bin
export PATH=$PATH:/usr/bin/verilator
export CALIPTRA_WORKSPACE=$W_ROOT/
export CALIPTRA_ROOT=$W_ROOT/caliptra-rtl-main/

Please let me know if i need provide any more details.

**Comment by howardtr** [comment]
In regards to (2), we have CI running with verilator

Here's a link to a github runner: https://github.com/chipsalliance/caliptra-rtl/actions/runs/5533741342/jobs/10097868907

Here's a link to the github runner commands: https://github.com/chipsalliance/caliptra-rtl/blob/main/.github/workflows/build-test-verilator.yml

As a first pass, can you run the smoke tests provided to make sure the environment is setup properly before trying to use other hex files?  Hopefully this will help ensure that the tooling and environment variables are all setup properly before trying to use different hex files

**Comment by ax-ag16** [comment]
Will try that Howard. 

We also another issue when integrating ICCM and DCCM  SRAMs to Caliptra.

1. Do we use the el2_mem_export interface from Caliptra top module port declaration or that port is only for tb.
2. If el2_mem_export is the one using which we should integrate the iccm and dccm memory can we refer the caliptra_veer_sram_export.sv as is to integrate the srams.

If point 2 is valid then on integrating the iccm and dccm sram in similar fashion we are getting the following error: 

`Error-[SE] Syntax error
  Following verilog source has syntax error : 
   /svc/test/riscv_core/veer_el2/rtl/el2_param.vh",
  18: token is 'pt'
  parameter el2_param_t pt = '{`
                          ^

**Comment by howardtr** [comment]
1.  Yes, use the el2_mem_export interface.  the [caliptra_top_tb_services.sv](https://github.com/chipsalliance/caliptra-rtl/blob/61fb5bd9eff399947f6e8621bdb1de03ff06e1b8/src/integration/tb/caliptra_top_tb_services.sv) has some functions that help preload all the hex files into the proper memory structures.  
The srams are here: https://github.com/chipsalliance/caliptra-rtl/blob/61fb5bd9eff399947f6e8621bdb1de03ff06e1b8/src/integration/tb/caliptra_top_tb_services.sv#L1040-L1043

The defines to all srams are here: https://github.com/chipsalliance/caliptra-rtl/blob/61fb5bd9eff399947f6e8621bdb1de03ff06e1b8/src/integration/tb/caliptra_top_tb_services.sv#L1225-L1232


Preloading is done here: https://github.com/chipsalliance/caliptra-rtl/blob/61fb5bd9eff399947f6e8621bdb1de03ff06e1b8/src/integration/tb/caliptra_top_tb_services.sv#L996-L998


2.  So that error looks familiar - I filed this a while back: https://github.com/chipsalliance/caliptra-rtl/issues/33, and this should have made it in: https://github.com/chipsalliance/caliptra-rtl/commit/a5fb21b3fde3e2f6c03e1a640e29b9a0bade7339

Can you confirm if you're using the latest from main?

Basically can you confirm if there is a `import el2_pkg::*;` prior to el2_param

See: https://github.com/chipsalliance/caliptra-rtl/blob/main/src/riscv_core/veer_el2/rtl/el2_veer.sv#L24-L26

**Comment by ax-ag16** [comment]
will double check that. Thanks

Can you please look into the issue [1] from the top. Regarding simv getting hung indefinitely. I have added the sim.log in the issue above.

**Comment by howardtr** [comment]
I'm not 100% familiar with that test, but from the logs I see:

```
No argument provided for CLP_MAX_CYCLES, defaulting to    20000000
DCCM pre-load from 50000000 to 5001ffff
DCCM pre-load completed
ICCM pre-load from 40000000 to 4001ffff
ICCM pre-load completed
MBOX pre-load from 00000000 to 00008000
MBOX pre-load completed

Warning-[STASKW_CO] Cannot open file
/User_data/home/agurram/Desktop/CALIPTRA/caliptra-rtl-main//src/integration/tb/caliptra_top_tb_services.sv, 529
  The file 'doe_test_vector.txt' could not be opened. No such file or 
  directory.
  Please ensure that the file exists with proper permissions.

```

Taking a look at caliptra_top_tb_services.sv, 529
https://github.com/chipsalliance/caliptra-rtl/blob/61fb5bd9eff399947f6e8621bdb1de03ff06e1b8/src/integration/tb/caliptra_top_tb_services.sv#L521-L529

This seems to imply that `python  doe_test_gen.py` needs to run to generate doe_test_vector.txt

Now, looking at @calebofearth comment in https://github.com/chipsalliance/caliptra-rtl/issues/120#issuecomment-1619126000 


> If you are running the simulation in a tool other than verilator, you should also copy [doe_test_gen.py](https://github.com/chipsalliance/caliptra-rtl/blob/dev-msft-iccm-lock-hex/src/doe/tb/doe_test_gen.py) and [ecdsa_secp384r1.exe](https://github.com/chipsalliance/caliptra-rtl/blob/dev-msft-iccm-lock-hex/src/ecc/tb/ecdsa_secp384r1.exe) into the test run directory.


I do see that you mentioned 

>  I have copied the ecdsa_secp384r1.exe and doe_test_gen.py files

@ax-ag16 : can you confirm where these files are copied to?  

It would also be good if you can add a line prior to the python script to check what it thinks the working directory is

```
+         $system("pwd"); 
         $system("python doe_test_gen.py"); 
```

If these directories are different, then move it to the one that pwd prints out and see what happens

**Comment by ax-ag16** [comment]
I pulled the latest RTL and I'm getting these syntax errors when running the simv.

**Error-[DPI-DIFNF] DPI import function not found
/User_data/home/agurram/Desktop/CALIPTRA/caliptra-hw//src/integration/test_suites/libs/jtagdpi/jtagdpi.sv, 34
  The definition of DPI import function/task 'jtagdpi_create' does not exist.
  Please check the stated DPI import function/task is defined, and its
  definition is either passed in a source file at compile-time, or provided in
  a shared library specified using the LRM Annex-J options at run-time.


Error-[DPI-DIFNF] DPI import function not found
/User_data/home/agurram/Desktop/CALIPTRA/caliptra-hw//src/integration/test_suites/libs/jtagdpi/jtagdpi.sv, 38
  The definition of DPI import function/task 'jtagdpi_close' does not exist.
  Please check the stated DPI import function/task is defined, and its
  definition is either passed in a source file at compile-time, or provided in
  a shared library specified using the LRM Annex-J options at run-time.**

Do I need any tool for use  DPI functions?

**Comment by howardtr** [comment]
@ax-ag16 - You'll need to pull in https://github.com/chipsalliance/caliptra-rtl/blob/61fb5bd9eff399947f6e8621bdb1de03ff06e1b8/src/integration/test_suites/libs/jtagdpi/jtagdpi.c

In parallel, @mkurc-ant - can you take a look at this?  

i see that jtagdpi is called out under verilator, but not for vcs

https://github.com/chipsalliance/caliptra-rtl/blob/61fb5bd9eff399947f6e8621bdb1de03ff06e1b8/tools/scripts/Makefile#L169-L171

**Comment by ax-ag16** [comment]
I pulled the filed you had mentioned but the libs/jtagdpi in test_suites has jtagdpi.c, jtagdpi.sv, jtagdpi.h files. Its still giving me the same error.

**Comment by howardtr** [comment]
I'm not an expert in dpi, but just double checking - did you make sure the object was built and linked properly?

@calebofearth - did you by chance run into this with internal regression testing?

**Comment by ax-ag16** [comment]
I already had the said jtagdpi.* files when i created the object files.

 I added the jtagdpi and tcp_server to COMP_LIB_NAMES in Makefile and when tried to recompile.. got another compile error that it couldn't find  

#include <svdpi.h>

Please let me know if i need any files. Thank you!

**Comment by calebofearth** [comment]
> I'm not an expert in dpi, but just double checking - did you make sure the object was built and linked properly?
> 
> @calebofearth - did you by chance run into this with internal regression testing?

I have not gotten the interactive JTAG debugger test (openOCD) working in our internal flow, but I was able to get it compiling on our internal VCS flow by adding:
https://github.com/chipsalliance/caliptra-rtl/blob/e40aae85b50bd54a732c65da1d8aa49c98d46369/src/integration/config/compile.yml#L77-L93

In the background, this produces a gcc command:

> gcc -m64 -fPIC -Wall -g  -Wno-unused-variable -Wno-unused-function -DVCS 
> -I/mypath/Caliptra/src/integration/test_suites/libs/tcp_server 
> -I/mypath/Caliptra/src/integration/test_suites/libs/jtagdpi 
> -I/mypath/synopsys/vcs-mx/2020.12-SP2-7/linux64/include 
> -shared 
> -c /mypath/Caliptra/src/integration/test_suites/libs/tcp_server/tcp_server.h 
> /mypath/Caliptra/src/integration/test_suites/libs/tcp_server/tcp_server.c 
> /mypath/Caliptra/src/integration/test_suites/libs/jtagdpi/jtagdpi.h 
> /mypath/Caliptra/src/integration/test_suites/libs/jtagdpi/jtagdpi.c
> 
> gcc -m64 -fPIC -shared   -o /mypath/scratch/cwhitehead/libs/vcs/R-2020.12-SP2-7_Full64/integration_lib/caliptra_top_tb/dpi/lib_dpi.so 
> /mypath/scratch/cwhitehead/libs/vcs/R-2020.12-SP2-7_Full64/integration_lib/caliptra_top_tb/dpi/tcp_server.o 
> /mypath/scratch/cwhitehead/libs/vcs/R-2020.12-SP2-7_Full64/integration_lib/caliptra_top_tb/dpi/jtagdpi.o
>

**Comment by ax-ag16** [comment]
@calebofearth 

The first error i got while generating simv is 

![image](https://github.com/chipsalliance/caliptra-rtl/assets/138075887/1f519d4b-2dcd-4cf1-986a-fe0022972e4e)

I commented out the IFDEF CALIPTRA_DEBUG_UNLOCKED for compiling purposes, but I got this error

![image](https://github.com/chipsalliance/caliptra-rtl/assets/138075887/8a32fbf9-37ac-4b6e-a982-895c12d7e71c)

Please let me know if there is any mode to set to get past this JTAG error?

**Comment by calebofearth** [comment]
The fix for your first issue was merged to main as part of https://github.com/chipsalliance/caliptra-rtl/pull/154 (commit https://github.com/chipsalliance/caliptra-rtl/commit/807de2b19b9a799e1063bdd3b96fcf480f473017 specifically has this fix).
For more context, take a look at https://github.com/chipsalliance/caliptra-rtl/issues/147

For your second issue, have you tried running the gcc commands I posted above to compile the DPI library?
If that does not work, you can always get the system to compile by just commenting out the jtagdpi:
https://github.com/chipsalliance/caliptra-rtl/blob/e40aae85b50bd54a732c65da1d8aa49c98d46369/src/integration/tb/caliptra_top_tb.sv#L1026-L1039

Eventually you should try to get the DPI running in your system though.

**Comment by ax-ag16** [comment]
I pulled the latest version and i am still facing the first issue.

![image](https://github.com/chipsalliance/caliptra-rtl/assets/138075887/80fbe108-a3fb-4870-a45c-dadbf4a1fead)

![image](https://github.com/chipsalliance/caliptra-rtl/assets/138075887/289b2a89-e0e4-4708-9d6d-ebf644e498e4)

Please let me know if i should be setting any mode to enable/disable DEBUG_UNLOCKED ?

**Comment by calebofearth** [comment]
Hi, I see what happened, the fix (made in MSFT internal repo) didn't sync back to the dev-msft branch correctly.
I have a fix pending: #158

**Comment by ax-ag16** [comment]
Hi, I am able to get past errors and compile simv but when run ./simv , my simulation still hangs after pre-loading DCCM, ICCM, MBOX files. I have attached the log file .
[sim.log](https://github.com/chipsalliance/caliptra-rtl/files/12100546/sim.log) . I am able to generate doe_test_vectors_all.txt and doe_test_vector.txt. Can you please help us get past this point?

**Comment by calebofearth** [comment]
This sim.log doesn't show much useful debug info - probably need to look at waves to diagnose.
Which firmware test is used to build program.hex for this run?
Is the clock alive?
Does the VeeR core come out of reset?
Are there any register accesses on the AHB bus?Instruction accesses on the IFU interface? Is there an NMI (You can find the mcause/mscause registers inside the VeeR hierarchy - what is their value?)

**Comment by calebofearth** [comment]
Are there any more updates on this? We will close the issue soon due to inactivity unless there are further questions.
